mlflow/models/signature.py

```diff
@@ -0,0 +1,449 @@
+(struct, map) are supported at this time.
+
+Input example specifies one or more valid inputs (observations) to the model. The example is stored
+as json and so the examples are limited to jsonable content with the exception of binary data
+provided the signature is defined. If the example includes binary data and model signature is
+defined, the binary data is base64 encoded before producing output.
+
+User can construct schema by hand or infer it from a dataset. Input examples are provided as is
+(pandas.DataFrame, numpy.ndarray, dictinary or list) and are json serialized  and stored as part of
+the model artifacts.
+"""
+
+import base64
+import importlib
+import os
+from enum import Enum
+
+import json
+import numpy as np
+import pandas as pd
+from typing import List, Dict, TypeVar
+
+from mlflow.exceptions import MlflowException
+from mlflow.utils.proto_json_utils import NumpyEncoder
+
+
+class TensorsNotSupportedException(MlflowException):
+    def __init__(self, msg):
+        super().__init__("Multidimensional arrays (aka tensors) are not supported. "
+                         "{}".format(msg))
+
+
+class DataType(Enum):
+    """
+    Data types supported by Mlflow model signature.
+    """
+
+    def __new__(cls, value, numpy_type):
+        res = object.__new__(cls)
+        res._value_ = value
+        res._numpy_type = numpy_type
+        return res
+
+    boolean = (1, np.bool)
+    integer = (2, np.int32)
+    long = (3, np.int64)
+    float = (4, np.float32)
+    double = (5, np.float64)
+    string = (6, np.str)
+    binary = (7, np.bytes_)
+
+    def __repr__(self):
+        return self.name
+
+    def to_numpy(self) -> np.dtype:
+        return self._numpy_type
+
+
+class ColSpec(object):
+    """
+    Specification of a column used in model signature.
+    Declares data type and optionally a name.
+    """
+
+    def __init__(self, name: str, type: DataType):  # pylint: disable=redefined-builtin
+        self.name = name
+        try:
+            self.type = getattr(DataType, type) if isinstance(type, str) else type
+        except AttributeError:
+            raise MlflowException("Unsupported type '{0}', expected instance of DataType or "
+                                  "one of {1}".format(type, [t.name for t in DataType]))
+
+    def to_dict(self) -> Dict[str, str]:
+        """
+        Serialize into a jsonable dictionary.
+        :return: dictionary representation of the column spec.
+        """
+        return {"name": self.name, "type": self.type.name}
+
+    def __eq__(self, other) -> bool:
+        names_eq = (self.name is None and other.name is None) or self.name == other.name
+        return names_eq and self.type == other.type
+
+    def __repr__(self) -> str:
+        return "{name}: {type}".format(name=self.name, type=self.type)
+
+
+class Schema(object):
+    """
+    Schema specifies column types (:py:class:`DataType`) in a dataset.
```

Comment:
Maybe document that there is one "input schema" and one "output schema" for a model, to help understandability and differentiating the concepts.